src/Illuminate/Database/Eloquent/HasCollection.php

```diff
@@ -15,6 +18,27 @@ trait HasCollection
      */
     public function newCollection(array $models = [])
     {
-        return new static::$collectionClass($models);
+        $collectionClass = $this->resolveCollectionAttribute() ?? static::$collectionClass;
```

Comment:
instead of `once()` this could also work, IMO easier to debug and troubleshoot.

```
static::$collectionClass ??= $this->resolveCollectionAttribute();

return new static::$collectionClass($models);
```